libtock/tock.c

```diff
@@ -35,22 +35,102 @@ int tock_enqueue(subscribe_upcall cb, int arg0, int arg1, int arg2, void* ud) {
+  if (command_return.type == TOCK_SYSCALL_SUCCESS) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE) {
+    return tock_status_to_returncode(command_return.data[0]);
+
+    // the remaining cases should not happen if using this function.
+
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U64) {
+    return RETURNCODE_FAIL;
+  }
+
+  // This should never happen as all syscall return variant cases have been
+  // handled.
+  return RETURNCODE_EBADRVAL;
+}
+
+int tock_command_return_u32_to_returncode(syscall_return_t command_return, uint32_t* val) {
+  if (command_return.type == TOCK_SYSCALL_SUCCESS_U32) {
+    *val = command_return.data[0];
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE) {
+    return tock_status_to_returncode(command_return.data[0]);
+
+    // the remaining cases should not happen if using this function.
+
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U64) {
+    return RETURNCODE_FAIL;
+  }
+
+  // This should never happen as all syscall return variant cases have been
+  // handled.
+  return RETURNCODE_EBADRVAL;
+}
+
+int tock_subscribe_return_to_returncode(subscribe_return_t subscribe_return) {
+  // If the subscribe was successful, easily return SUCCESS.
+  if (subscribe_return.success) {
+    return RETURNCODE_SUCCESS;
+  } else {
+    // Not success, so return the proper returncode.
+    return tock_status_to_returncode(subscribe_return.status);
+  }
+}
+
+int tock_allow_rw_return_to_returncode(allow_rw_return_t allow_return) {
```

Comment:
in general, I think all these helper functions are fantastic and something we should have done from the start.